README.md
# emis-alert

[![Build Status](https://travis-ci.org/CodeTanzania/emis-alert.svg?branch=develop)](https://travis-ci.org/CodeTanzania/emis-alert)
[![Dependencies Status](https://david-dm.org/CodeTanzania/emis-alert/status.svg?style=flat-square)](https://david-dm.org/CodeTanzania/emis-alert)
[![Deploy to Heroku](https://www.herokucdn.com/deploy/button.png)](https://heroku.com/deploy?template=https://github.com/CodeTanzania/emis-alert/tree/develop)

A representation of an envelope(or payload) which carries disaster notifications(or warning) from source(s) to audience(s).

[Demo](https://emis-alert.herokuapp.com/v1/alerts)

## Domain Model

![EMIS Alert Domain Model](https://raw.githubusercontent.com/CodeTanzania/emis-alert/develop/specifications/alert.model.png)

## Requirements

- [NodeJS v8.11.1+](https://nodejs.org)
- [Npm v5.6.0+](https://www.npmjs.com/)
- [MongoDB v3.4.10+](https://www.mongodb.com/)
- [Mongoose v5.1.2+](https://github.com/Automattic/mongoose)

## Installation

```sh
npm install @codetanzania/emis-alert --save
```

## Usage

```js
const { app } = require('@codetanzania/emis-alert');
app.start((error) => { ... });
```

## References
- [Common Alerting Protocol](https://en.wikipedia.org/wiki/Common_Alerting_Protocol)
- [Common Alerting Protocol Version 1.2](http://docs.oasis-open.org/emergency/cap/v1.2/CAP-v1.2-os.html)
- [CAP-v1.2-PR03](http://docs.oasis-open.org/emergency/cap/v1.2/pr03/CAP-v1.2-PR03.pdf)
- [Google Public Alerts CAP v1.0](https://developers.google.com/public-alerts/reference/cap-google)
- [Guidelines for Implementation CAP](https://library.wmo.int/pmb_ged/wmo_1109_en.pdf)
- [Wikipedia: ISO 22324](https://en.wikipedia.org/wiki/ISO_22324)

## Testing

- Clone this repository

- Install all development dependencies

```sh
npm install
```

- Run example

```sh
npm run dev
```

- Then run test

```sh
npm test
```

## Contribute

It will be nice, if you open an issue first so that we can know what is going on, then, fork this repo and push in your ideas. Do not forget to add a bit of test(s) of what value you adding.

## Licence

The MIT License (MIT)

Copyright (c) 2018 CodeTanzania & Contributors

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
#### 1.5.0 (2019-06-20)

##### Chores

* **deps:**  force latest version & audit fix ([98b8d1a2](https://github.com/CodeTanzania/emis-alert/commit/98b8d1a2d892f23add7722efaf72d38208e53c68))

##### New Features

*  add agency and locations ([6990c827](https://github.com/CodeTanzania/emis-alert/commit/6990c827f9aed23e0a40c2503cab0bf0b8e76d15))
*  add agency and locations fields ([ad418d41](https://github.com/CodeTanzania/emis-alert/commit/ad418d41d731f161e744fad514b2b638f0fe698c))
*  refactor to use campaign on alert sending ([1f3c27d1](https://github.com/CodeTanzania/emis-alert/commit/1f3c27d16cdd02ec3e4b639c1b80f265ebf0568d))

##### Code Style Changes

*  fix unused imports ([2fe7877a](https://github.com/CodeTanzania/emis-alert/commit/2fe7877ac9086a90af0084e2d3063885b71f7148))

#### 1.4.1 (2019-06-17)

##### Chores

* **package.json:**  update package version to 1.4.1 ([a0ea96cf](https://github.com/CodeTanzania/emis-alert/commit/a0ea96cfc4a7e01bff4d740786e7bb1184b38cf2))
* **.gitignore:**  ignore .idea and .vscode configuration folders ([3dbace88](https://github.com/CodeTanzania/emis-alert/commit/3dbace88f4e8e26449c08d0898fe0f9a9a38f31c))
* **deps:**  force latest version & audit fix ([208a6041](https://github.com/CodeTanzania/emis-alert/commit/208a60419e9a73de463f5ad40a448b5061056380))
*  ignore neovim editor configs ([7051e424](https://github.com/CodeTanzania/emis-alert/commit/7051e424632531df39f4b628dbf6fade227a115b))
*  remove unused jsbeautify file ([03e539a5](https://github.com/CodeTanzania/emis-alert/commit/03e539a5fda3dcf0158da2ff80d9dc3650463748))
*  install and configure prettier ([48d482a0](https://github.com/CodeTanzania/emis-alert/commit/48d482a0ee184e093500fe3e87c1d4033e5a48fe))

##### Bug Fixes

* **seed:**  use prepareSeedCriteria to seed data ([878553cb](https://github.com/CodeTanzania/emis-alert/commit/878553cb11d7d52e346110347f6ce57d9f54da9e))

##### Code Style Changes

*  format files with prettier ([a75c6c01](https://github.com/CodeTanzania/emis-alert/commit/a75c6c01a511f63466fea523526e50a4955051a3))

#### 1.4.0 (2019-05-30)

##### Chores

* **deps:**  force latest version & audit fix ([f05d71aa](https://github.com/CodeTanzania/emis-alert/commit/f05d71aaee752fc5d80d66372abd6e91155a7f4b))
*  force latest dependencies ([158ef975](https://github.com/CodeTanzania/emis-alert/commit/158ef975a1ef42ceb08b89f4662d95202de74da2))

##### New Features

*  format alert dates on exports and reorder columns ([77328ee9](https://github.com/CodeTanzania/emis-alert/commit/77328ee905041cc8e08e7ad4550c1d479f807480))
*  add export colum definition on Alert & Alert Source ([e88bd4ac](https://github.com/CodeTanzania/emis-alert/commit/e88bd4ac5e0e567f12f5ef82c8667369b6b0b53e))
*  apply exprtable plugin to Alert and AlertSource models ([d7aa41a5](https://github.com/CodeTanzania/emis-alert/commit/d7aa41a5f725db2436153f9b42bd0003b03b70df))

##### Bug Fixes

*  update exportable plugins to remove unclered exportable fields ([4c09ef7c](https://github.com/CodeTanzania/emis-alert/commit/4c09ef7c3ee5b7d042bbcf97eab59c1c18ba8d68))

##### Refactors

* **deps:**  update to use latest dependencies api ([f5b05064](https://github.com/CodeTanzania/emis-alert/commit/f5b0506476861508670aed0b21f8e28ce946ea86))

#### 1.3.0 (2019-01-29)

##### New Features

*  implement alert source router ([6c674399](https://github.com/CodeTanzania/emis-alert/commit/6c6743999de31bb0d842552b45405a6eb0f3996c))
*  implement alert source model ([0506ec59](https://github.com/CodeTanzania/emis-alert/commit/0506ec593f9d7e7b3903411edc602f7cd5705316))

# v1.2.2 - 2018-12-24
- Build latest apidoc
- Force latest dependencies & audit fix
- Refactor and improve specs readability

# v1.2.1 - 2018-12-17
- Build latest apidoc
- Reorganize dependencies
- Apply taggable plugin
- Force latest dependencies & audit fix
- Improve package main jsdoc
- Improve example app

# v1.2.0 - 2018-11-24  
- Wire mongoose json schema to extract json schema from alert schema     
- Add alerts/schema endpoint to expose Alert json schema
- Force latest dependencies
- Build latest apidoc

# v1.1.0 - 2018-11-13
- Build latest API doc
- Add notification worker on example
- Add SMTP configuration on example
- Force latest dependencies
- Wire notification queue on alert afterPost
- Re-organize dependencies & add redis in travis build
- Fix centroid calculations
- Add real seed on example
- Add Alert upsert logic
- Add Alert seed logic
- Add Alert upsert integration test
- Add Alert seed integration test
- Add color on alert schema
- Derive alert color based on severity
- Improve alert router api doc
- Improve model unit spec



# v1.0.0 - 2018-10-01
- Initial release


lib/alert.http.router.js
'use strict';

/**
 * @apiDefine Alert  Alert
 *
 * @apiDescription A representation of an envelope(or payload) which carries
 * disaster notifications(or warning) from source(s) to audience(s).
 *
 * @see {@link http://docs.oasis-open.org/emergency/cap/v1.2/CAP-v1.2-os.html}
 * @see {@link http://docs.oasis-open.org/emergency/cap/v1.2/pr03/CAP-v1.2-PR03.pdf}
 * @see {@link https://en.wikipedia.org/wiki/Common_Alerting_Protocol}
 * @see {@link https://developers.google.com/public-alerts/reference/cap-google}
 * @see {@link https://library.wmo.int/pmb_ged/wmo_1109_en.pdf}
 * @see {@link https://en.wikipedia.org/wiki/ISO_22324}
 *
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @public
 */

/**
 * @apiDefine Alert
 * @apiSuccess {String} _id Unique alert identifier
 * @apiSuccess {String} source Human readable name of party(agency, authority etc)
 * issuing(or generating) an alert.
 * @apiSuccess {String} [number] Human readable, system specific identifier of
 * an alert.
 * @apiSuccess {String} [status] Human readable code denoting the appropriate
 * handling of an alert.
 * @apiSuccess {String} [type] Human readable code denoting the nature
 * of an alert.
 * @apiSuccess {String} [scope] Human readable code denoting the intended
 * distribution of an alert.
 * @apiSuccess {String} [category] Human readable code denoting the category
 * of an alert.
 * @apiSuccess {String} event Human readable text denoting the subject of
 * an alert.
 * @apiSuccess {String} [response] Human readable code denoting the type of
 * action recommended for the target audience as per instruction of an alert.
 * @apiSuccess {String} [urgency] Human readable code denoting the urgency
 * of an alert.
 * @apiSuccess {String} [severity] Human readable code denoting the severity
 * of an alert.
 * @apiSuccess {String} [certainty] Human readable code denoting the certainty
 * of an alert.
 * @apiSuccess {String} [headline] A brief human-readable headline(title) of
 * an alert.
 * @apiSuccess {String} [description] An extended human readable description
 * of an alert.
 * @apiSuccess {String} [instruction] Human readable text describing the
 * recommended action to be taken by recipients of an alert.
 * @apiSuccess {String} [url] A full, absolute URI for an HTML page or other
 * text resource with additional or reference information regarding an alert.
 * @apiSuccess {String} area Human readble text describing the affected area(s)
 * of an alert.
 * @apiSuccess {Geometry} [geometry] The paired values of geo-points defining
 * a geometry(i.e polygon, line etc) that delineates the affected area(s) of
 * an alert.
 * @apiSuccess {Point} [centroid] The centre of affected area of an alert.
 * @apiSuccess {Date} [reportedAt] Date and time when an alert was
 * issued(or generated) by sender(source, generator).
 * @apiSuccess {Date} [expectedAt] The expected time of the beginning of
 * an alert.
 * @apiSuccess {Date} [expiredAt] The expiry time of the information
 * of an alert.
 * @apiSuccess {String} [direction=Inbound]  Human readable direction(received
 * or given) of the alert.
 * @apiSuccess {String} [color] A color code(in hexadecimal format) used to
 * differentiate an alert visually.
 * @apiSuccess {Date} createdAt Date when alert was created.
 * @apiSuccess {Date} updatedAt Date when alert was last updated.
 */

/**
 * @apiDefine Alerts
 * @apiSuccess {Object[]} data List of alerts
 * @apiSuccess {String} data._id Unique alert identifier
 * @apiSuccess {String} data.source Human readable name of party(agency,
 * authority etc) issuing(or generating) an alert.
 * @apiSuccess {String} [data.number] Human readable, system specific identifier
 * of an alert.
 * @apiSuccess {String} [data.status] Human readable code denoting the
 * appropriate handling of an alert.
 * @apiSuccess {String} [data.type] Human readable code denoting the nature
 * of an alert.
 * @apiSuccess {String} [data.scope] Human readable code denoting the intended
 * distribution of an alert.
 * @apiSuccess {String} [data.category] Human readable code denoting the category
 * of an alert.
 * @apiSuccess {String} data.event Human readable text denoting the subject of
 * an alert.
 * @apiSuccess {String} [data.response] Human readable code denoting the type of
 * action recommended for the target audience as per instruction of an alert.
 * @apiSuccess {String} [urgency] Human readable code denoting the urgency
 * of an alert.
 * @apiSuccess {String} [data.severity] Human readable code denoting the severity
 * of an alert.
 * @apiSuccess {String} [data.certainty] Human readable code denoting the certainty
 * of an alert.
 * @apiSuccess {String} [data.headline] A brief human-readable headline(title) of
 * an alert.
 * @apiSuccess {String} [data.description] An extended human readable description
 * of an alert.
 * @apiSuccess {String} [data.instruction] Human readable text describing the
 * recommended action to be taken by recipients of an alert.
 * @apiSuccess {String} [data.url] A full, absolute URI for an HTML page or other
 * text resource with additional or reference information regarding an alert.
 * @apiSuccess {String} data.area Human readble text describing the affected area(s)
 * of an alert.
 * @apiSuccess {Geometry} [data.geometry] The paired values of geo-points defining
 * a geometry(i.e polygon, line etc) that delineates the affected area(s) of
 * an alert.
 * @apiSuccess {Point} [data.centroid] The centre of affected area of an alert.
 * @apiSuccess {Date} [data.reportedAt] Date and time when an alert was
 * issued(or generated) by sender(source, generator).
 * @apiSuccess {Date} [data.expectedAt] The expected time of the beginning of
 * an alert.
 * @apiSuccess {Date} [data.expiredAt] The expiry time of the information
 * of an alert.
 * @apiSuccess {String} [data.direction=Inbound]  Human readable direction(received
 * or given) of the alert.
 * @apiSuccess {String} [data.color] A color code(in hexadecimal format) used to
 * differentiate an alert visually.
 * @apiSuccess {Date} data.createdAt Date when alert was created.
 * @apiSuccess {Date} data.updatedAt Date when alert was last updated.
 * @apiSuccess {Number} total Total number of alert
 * @apiSuccess {Number} size Number of alert returned
 * @apiSuccess {Number} limit Query limit used
 * @apiSuccess {Number} skip Query skip/offset used
 * @apiSuccess {Number} page Page number
 * @apiSuccess {Number} pages Total number of pages
 * @apiSuccess {Date} lastModified Date and time at which latest alert
 * was last modified
 */

/**
 * @apiDefine AlertSuccessResponse
 * @apiSuccessExample {json} Success-Response:
 * {
 *    "_id": "5be1df42842d9d5bf8556da8",
 *    "sender": "Vandervort",
 *    "number": "55DAEB99-14C5-4009-9A84-EBF01D96D89C",
 *    "status": "Exercise",
 *    "type": "Error",
 *    "scope": "Public",
 *    "category": "Fire",
 *    "event": "Omnis et delectus id omnis quibusdam.",
 *    "response": "AllClear",
 *    "urgency": "Past",
 *    "severity": "Severe",
 *    "certainty": "Observed",
 *    "headline": "Autem sint hic dolorem quasi aliquam amet beatae.",
 *    "description": "Ut voluptate quo voluptatem ea asperiores.",
 *    "instruction": "Et doloremque odit alias optio nihil temporibus soluta.",
 *    "url": "https://loyce.org",
 *    "area": "Avon",
 *    "geometry":
 *    {
 *        "type": "Polygon",
 *        "coordinates": [
 *            [
 *                [34.0962460876592, -6.027472074582597],
 *                [36.14211410064278, -7.73276252407186],
 *                [34.26678374339001, -11.052568169309106],
 *                [37.9984237359413, -19.518211005210446],
 *                [34.81395505104936, -16.756448591687228],
 *                [33.1875973147786, -19.028057614810873],
 *                [33.19738425670787, -11.327556097253908],
 *                [29.532252395715762, -9.526871069904153],
 *                [27.14333524479317, -5.2808693952369214],
 *                [33.662964128882464, -10.049782036387153],
 *                [34.0962460876592, -6.027472074582597]
 *            ]
 *        ]
 *    },
 *    "centroid":
 *    {
 *        "type": "Point",
 *        "coordinates": [39.14056019762375, -3.7587456547395846]
 *    },
 *    "reportedAt": "2018-09-30T05:05:28.689Z",
 *    "expectedAt": "2018-10-29T11:29:16.562Z",
 *    "expiredAt": "2018-10-25T05:35:26.462Z",
 *    "direction": "Outbound",
 *    "color": "#FE9901",
 *    "updatedAt": "2018-11-06T18:36:50.998Z",
 *    "createdAt": "2018-11-06T18:36:50.998Z"
 * }
 */

/**
 * @apiDefine AlertsSuccessResponse
 * @apiSuccessExample {json} Success-Response:
 * {
 *    "data": [
 *    {
 *        "sender": "Vandervort",
 *        "number": "55DAEB99-14C5-4009-9A84-EBF01D96D89C",
 *        "status": "Exercise",
 *        "type": "Error",
 *        "scope": "Public",
 *        "category": "Fire",
 *        "event": "Omnis et delectus id omnis quibusdam.",
 *        "response": "AllClear",
 *        "urgency": "Past",
 *        "severity": "Severe",
 *        "certainty": "Observed",
 *        "headline": "Autem sint hic dolorem quasi aliquam amet beatae.",
 *        "description": "Ut voluptate quo voluptatem ea asperiores et.",
 *        "instruction": "Et doloremque odit alias optio nihil temporibus.",
 *        "url": "https://loyce.org",
 *        "area": "Avon",
 *        "geometry":
 *        {
 *            "type": "Polygon",
 *            "coordinates": [
 *                [
 *                    [34.0962460876592, -6.027472074582597],
 *                    [36.14211410064278, -7.73276252407186],
 *                    [34.26678374339001, -11.052568169309106],
 *                    [37.9984237359413, -19.518211005210446],
 *                    [34.81395505104936, -16.756448591687228],
 *                    [33.1875973147786, -19.028057614810873],
 *                    [33.19738425670787, -11.327556097253908],
 *                    [29.532252395715762, -9.526871069904153],
 *                    [27.14333524479317, -5.2808693952369214],
 *                    [33.662964128882464, -10.049782036387153],
 *                    [34.0962460876592, -6.027472074582597]
 *                ]
 *            ]
 *        },
 *        "centroid":
 *        {
 *            "type": "Point",
 *            "coordinates": [39.14056019762375, -3.7587456547395846]
 *        },
 *        "reportedAt": "2018-09-30T05:05:28.689Z",
 *        "expectedAt": "2018-10-29T11:29:16.562Z",
 *        "expiredAt": "2018-10-25T05:35:26.462Z",
 *        "direction": "Outbound",
 *        "color": "#FE9901",
 *        "_id": "5be1df42842d9d5bf8556da8",
 *        "updatedAt": "2018-11-06T18:36:50.998Z",
 *        "createdAt": "2018-11-06T18:36:50.998Z"
 *    }],
 *    "total": 32,
 *    "size": 10,
 *    "limit": 10,
 *    "skip": 0,
 *    "page": 1,
 *    "pages": 32,
 *    "lastModified": "2018-11-06T18:36:50.999Z"
 * }
 */

/* dependencies */
const _ = require('lodash');
const { getString } = require('@lykmapipo/env');
const { include } = require('@lykmapipo/include');
const Router = require('@lykmapipo/express-common').Router;

/* local constants */
const API_VERSION = getString('API_VERSION', '1.0.0');
const PATH_LIST = '/alerts';
const PATH_SINGLE = '/alerts/:id';
const PATH_SCHEMA = '/alerts/schema/';
const PATH_EXPORT = '/alerts/export';

/* declarations */
const Alert = include(__dirname, 'alert.model');
const router = new Router({
  version: API_VERSION,
});

/**
 * @api {get} /alerts List Alerts
 * @apiVersion 1.0.0
 * @apiName GetAlerts
 * @apiGroup Alert
 * @apiDescription Returns a list of alerts
 * @apiUse RequestHeaders
 * @apiUse Alerts
 *
 *
 * @apiUse RequestHeadersExample
 * @apiUse AlertsSuccessResponse
 * @apiUse JWTError
 * @apiUse JWTErrorExample
 * @apiUse AuthorizationHeaderError
 * @apiUse AuthorizationHeaderErrorExample
 */
router.get(PATH_LIST, function getAlerts(request, response, next) {
  //obtain request options
  const options = _.merge({}, request.mquery);

  Alert.get(options, function onGetAlerts(error, results) {
    //forward error
    if (error) {
      next(error);
    }

    //handle response
    else {
      response.status(200);
      response.json(results);
    }
  });
});

/**
 * @api {get} /alerts/schema Get Alerts Schema
 * @apiVersion 1.0.0
 * @apiName GetAlertSchema
 * @apiGroup Alert
 * @apiDescription Returns alert json schema definition
 * @apiUse RequestHeaders
 */
router.get(PATH_SCHEMA, function getAlertSchema(request, response) {
  const schema = Alert.jsonSchema();
  response.status(200);
  response.json(schema);
});

/**
 * @api {get} /alerts/export Export Alerts
 * @apiVersion 1.0.0
 * @apiName ExportAlerts
 * @apiGroup Alert
 * @apiDescription Export alerts as csv
 * @apiUse RequestHeaders
 */
router.get(PATH_EXPORT, function getExport(request, response) {
  //obtain request options
  const options = _.merge({}, request.mquery);

  // reply with csv export
  response.attachment(`alerts_exports_${Date.now()}.csv`);
  response.status(200);
  Alert.exportCsv(options).pipe(response);
});

/**
 * @api {post} /alerts Create New Alert
 * @apiVersion 1.0.0
 * @apiName PostAlert
 * @apiGroup Alert
 * @apiDescription Create new alert
 * @apiUse RequestHeaders
 * @apiUse Alert
 *
 *
 * @apiUse RequestHeadersExample
 * @apiUse AlertSuccessResponse
 * @apiUse JWTError
 * @apiUse JWTErrorExample
 * @apiUse AuthorizationHeaderError
 * @apiUse AuthorizationHeaderErrorExample
 */
router.post(PATH_LIST, function postAlert(request, response, next) {
  //obtain request body
  const body = _.merge({}, request.body);

  Alert.post(body, function onPostAlert(error, created) {
    //forward error
    if (error) {
      next(error);
    }

    //handle response
    else {
      response.status(201);
      response.json(created);
    }
  });
});

/**
 * @api {get} /alerts/:id Get Existing Alert
 * @apiVersion 1.0.0
 * @apiName GetAlert
 * @apiGroup Alert
 * @apiDescription Get existing alert
 * @apiUse RequestHeaders
 * @apiUse Alert
 *
 *
 * @apiUse RequestHeadersExample
 * @apiUse AlertSuccessResponse
 * @apiUse JWTError
 * @apiUse JWTErrorExample
 * @apiUse AuthorizationHeaderError
 * @apiUse AuthorizationHeaderErrorExample
 */
router.get(PATH_SINGLE, function getAlert(request, response, next) {
  //obtain request options
  const options = _.merge({}, request.mquery);

  //obtain alert id
  options._id = request.params.id;

  Alert.getById(options, function onGetAlert(error, found) {
    //forward error
    if (error) {
      next(error);
    }

    //handle response
    else {
      response.status(200);
      response.json(found);
    }
  });
});

/**
 * @api {patch} /alerts/:id Patch Existing Alert
 * @apiVersion 1.0.0
 * @apiName PatchAlert
 * @apiGroup Alert
 * @apiDescription Patch existing alert
 * @apiUse RequestHeaders
 * @apiUse Alert
 *
 *
 * @apiUse RequestHeadersExample
 * @apiUse AlertSuccessResponse
 * @apiUse JWTError
 * @apiUse JWTErrorExample
 * @apiUse AuthorizationHeaderError
 * @apiUse AuthorizationHeaderErrorExample
 */
router.patch(PATH_SINGLE, function patchAlert(request, response, next) {
  //obtain alert id
  const { id } = request.params;

  //obtain request body
  const patches = _.merge({}, request.body);

  Alert.patch(id, patches, function onPatchAlert(error, patched) {
    //forward error
    if (error) {
      next(error);
    }

    //handle response
    else {
      response.status(200);
      response.json(patched);
    }
  });
});

/**
 * @api {put} /alerts/:id Put Existing Alert
 * @apiVersion 1.0.0
 * @apiName PutAlert
 * @apiGroup Alert
 * @apiDescription Put existing alert
 * @apiUse RequestHeaders
 * @apiUse Alert
 *
 *
 * @apiUse RequestHeadersExample
 * @apiUse AlertSuccessResponse
 * @apiUse JWTError
 * @apiUse JWTErrorExample
 * @apiUse AuthorizationHeaderError
 * @apiUse AuthorizationHeaderErrorExample
 */
router.put(PATH_SINGLE, function putAlert(request, response, next) {
  //obtain alert id
  const { id } = request.params;

  //obtain request body
  const updates = _.merge({}, request.body);

  Alert.put(id, updates, function onPutAlert(error, updated) {
    //forward error
    if (error) {
      next(error);
    }

    //handle response
    else {
      response.status(200);
      response.json(updated);
    }
  });
});

/**
 * @api {delete} /alerts/:id Delete Existing Alert
 * @apiVersion 1.0.0
 * @apiName DeleteAlert
 * @apiGroup Alert
 * @apiDescription Delete existing alert
 * @apiUse RequestHeaders
 * @apiUse Alert
 *
 *
 * @apiUse RequestHeadersExample
 * @apiUse AlertSuccessResponse
 * @apiUse JWTError
 * @apiUse JWTErrorExample
 * @apiUse AuthorizationHeaderError
 * @apiUse AuthorizationHeaderErrorExample
 */
router.delete(PATH_SINGLE, function deleteAlert(request, response, next) {
  //obtain alert id
  const { id } = request.params;

  Alert.del(id, function onDeleteAlert(error, deleted) {
    //forward error
    if (error) {
      next(error);
    }

    //handle response
    else {
      response.status(200);
      response.json(deleted);
    }
  });
});

/* expose router */
module.exports = router;
